### PR TITLE
Raise EOF if response body content length is known

### PR DESCRIPTION
**Pull request by joshcooper**
If response body content length is known, then read the complete
response body or raise EOF if it is truncated.

[[Bug #14972][ruby-core:88324]](https://bugs.ruby-lang.org/issues/14972)

### CODE REVIEW

File: lib/net/http/response.rb [hunk 1]
```diff
@@ -290,7 +290,7 @@ def read_body_0(dest)
 
       clen = content_length()
       if clen
-        @socket.read clen, dest, true   # ignore EOF
+        @socket.read clen, dest
```
Comment (joshcooper) [inline lib/net/http/response.rb]: I'm assuming `@socket.read` semantics are as described in https://ruby-doc.org/core-2.5.0/IO.html#method-i-read, specifically "Note that this method behaves like the fread() function in C"

Comment (ioquatix) [reply lib/net/http/response.rb]: Uh, `#read` doesn't accept 3 arguments? I'm a bit confused.

Comment (joshcooper) [reply lib/net/http/response.rb]: It ends up calling [`Net::BufferedIO#read`](https://github.com/ruby/ruby/blob/v2_5_3/lib/net/protocol.rb#L117) which accepts the optional `ignore_eof` parameter.

**Comment by joshcooper** [comment]
Per https://tools.ietf.org/html/rfc7230#section-3.4:

>  A message that uses a valid Content-Length is incomplete
>   if the size of the message body received (in octets) is less than the
>  value given by Content-Length.

**Comment by jeremyevans** [comment]
The net-http gem now has its own repository, so this should be closed. I added a pull request to that repository that offers a way to support this in a backwards compatible manner, as requested by @nurse: https://github.com/ruby/net-http/pull/15.

**Comment by joshcooper** [comment]
Thanks @jeremyevans!

**Pull request closed without merging**